src/main/__test__/context-handler.test.ts

```diff
@@ -0,0 +1,114 @@
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
+ * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
+ * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
+ * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+import { UserStore } from "../../common/user-store";
+import { ContextHandler } from "../context-handler";
+import { PrometheusProvider, PrometheusProviderRegistry, PrometheusService } from "../prometheus";
+
+enum ServiceResult {
+  Success,
+  Failure,
+  Undefined,
+}
+
+class TestProvider extends PrometheusProvider {
+  name = "TestProvider1";
+  rateAccuracy = "1h";
+  isConfigurable = false;
+
+  constructor(public id: string, public alwaysFail: ServiceResult) {
+    super();
+  }
+
+  getQuery(): string {
+    throw new Error("getQuery is not implemented.");
+  }
+
+  async getPrometheusService(): Promise<PrometheusService> {
+    switch (this.alwaysFail) {
+      case ServiceResult.Success:
+        return {
+          id: this.id,
+          namespace: "default",
+          port: 7000,
+          service: "",
+        };
+      case ServiceResult.Failure:
+        throw new Error("does fail");
+      case ServiceResult.Undefined:
+        return undefined;
+    }
+  }
+}
+
+function getHandler() {
+  return new ContextHandler(({
+    getProxyKubeconfig: (): any => ({
+      makeApiClient: (): any => undefined,
+    }),
+    apiUrl: "http://localhost:81",
+  }) as any);
+}
+
+describe("ContextHandler", () => {
+  beforeEach(() => {
+    PrometheusProviderRegistry.createInstance();
+    UserStore.createInstance();
+  });
+
+  afterEach(() => {
+    PrometheusProviderRegistry.resetInstance();
+    UserStore.resetInstance();
+  });
+
+  describe("getPrometheusService", () => {
+    it("should pick the first provider of 1 success", async () => {
+      PrometheusProviderRegistry.getInstance()
+        .registerProvider(new TestProvider("id_foobar", ServiceResult.Success));
+
+      const service = await getHandler().getPrometheusService();
+
+      expect(service.id === "id_foobar");
+    });
+
+    it("should pick the first provider of 1 success, after 2 fails", async () => {
+      PrometheusProviderRegistry.getInstance()
+        .registerProvider(new TestProvider("id_foobar_1", ServiceResult.Failure))
+        .registerProvider(new TestProvider("id_foobar_2", ServiceResult.Undefined))
+        .registerProvider(new TestProvider("id_foobar_3", ServiceResult.Success));
+
+      const service = await getHandler().getPrometheusService();
+
+      expect(service.id === "id_foobar_3");
```

Comment:
Before you corrected this test the first one that succeeded would've been "id_foobar_2", so this line should've failed, right?

Reply:
Oh right, I thought you were talking about the third test. Will investigate.

Reply:
I don't know why this test previously was passing. But I have added many more tests, and even one to validate that that specific case doesn't happen.